src/dma.rs

```diff
@@ -0,0 +1,699 @@
+macro_rules! dma {
+    (
+        $DMAx:ident, $dmax:ident, $dmaxen:ident,
+        channels: {
+            $( $Ci:ident: (
+                $chi:ident,
+                $htifi:ident, $tcifi:ident, $teifi:ident, $gifi:ident,
+                $chtifi:ident, $ctcifi:ident, $cteifi:ident, $cgifi:ident
+            ), )+
+        },
+    ) => {
+        pub mod $dmax {
+            use super::*;
+            use crate::stm32::$DMAx;
+
+            impl DmaExt for $DMAx {
+                type Channels = Channels;
+
+                fn split(self, ahb: &mut AHB) -> Channels {
+                    ahb.enr().modify(|_, w| w.$dmaxen().set_bit());
+
+                    let mut channels = Channels {
+                        $( $chi: $Ci { _0: () }, )+
+                    };
+
+                    channels.reset();
+                    channels
+                }
+            }
+
+            /// DMA channels
+            pub struct Channels {
+                $( pub $chi: $Ci, )+
+            }
+
+            impl Channels {
+                /// Reset the control registers of all channels.
+                /// This stops any ongoing transfers.
+                fn reset(&mut self) {
+                    $( self.$chi.reset(); )+
+                }
+            }
+
+            $(
+                /// Singleton that represents a DMA channel
+                pub struct $Ci {
+                    _0: (),
+                }
+
+                impl private::Channel for $Ci {
+                    fn ch(&self) -> &stm32::dma1::CH {
+                        // NOTE(unsafe) $Ci grants exclusive access to this register
+                        unsafe { &(*$DMAx::ptr()).$chi }
+                    }
+                }
+
+                impl Channel for $Ci {
+                    fn event_occured(&self, event: Event) -> bool {
+                        use Event::*;
+
+                        // NOTE(unsafe) atomic read
+                        let flags = unsafe { (*$DMAx::ptr()).isr.read() };
+                        match event {
+                            HalfTransfer => flags.$htifi().bit_is_set(),
+                            TransferComplete => flags.$tcifi().bit_is_set(),
+                            TransferError => flags.$teifi().bit_is_set(),
+                            Any => flags.$gifi().bit_is_set(),
+                        }
+                    }
+
+                    fn clear_event(&mut self, event: Event) {
+                        use Event::*;
+
+                        // NOTE(unsafe) atomic write to a stateless register
+                        unsafe {
+                            &(*$DMAx::ptr()).ifcr.write(|w| match event {
+                                HalfTransfer => w.$chtifi().set_bit(),
+                                TransferComplete => w.$ctcifi().set_bit(),
+                                TransferError => w.$cteifi().set_bit(),
+                                Any => w.$cgifi().set_bit(),
+                            });
```

Comment:
I tend to agree. But what about the scenario where both `TransferComplete` and `HalfTransfer` occured and the user clears `TransferComplete`? In this case, wouldn't they expect a call with `Any` to still return `true`, since the `HalfTransfer` is still set?

IMO, we should only clear the global flag when all other flags are cleared too. Or just leave it as is and document that the user has to take care of clearing the global flag.

Reply:
Hmm, makes sense, I think the best solution is to leave as is and have a good documentation about this then.